### PR TITLE
Mantid calibration for Mandi as good as that from ISAW

### LINKED ISSUE #15644: SCDCallibratePanels for MANDI
Two issues:
1. Calibration does not do as well as ISAW
2. Output is "clunky"

### PR DESCRIPTION
**Pull request by VickieLynch**
Description of work.
Two problems were found in SCD calibrations in Mantid.  First the B matrix was not fixed and second there were some errors in the analytic derivatives.  It now uses numerical derivatives and only the U matrix is allowed to change.

**To test:**

Get mandi input files from dropbox and use python script to check that indexing error from Mantid calibration is as good as one from ISAW.
https://www.dropbox.com/s/k6f6auy4ghx0p4q/natrolite_Niggli_0.12.peaks?dl=0
https://www.dropbox.com/s/3r0n6u5qa16gqbb/natrolite_Niggli.mat?dl=0
https://www.dropbox.com/s/tg8vklot51cap85/natrolite_Niggli_0.12_2016.peaks?dl=0

``` python
import time

LoadIsawPeaks(Filename='natrolite_Niggli_0.12.peaks', OutputWorkspace='natrolite')
LoadIsawUB(InputWorkspace='natrolite', Filename='natrolite_Niggli.mat')
CloneWorkspace(InputWorkspace="natrolite",OutputWorkspace="natrolite_orig")
CloneWorkspace(InputWorkspace="natrolite_orig",OutputWorkspace="natrolite_Simplex")
num,err=IndexPeaks(PeaksWorkspace='natrolite_orig',RoundHKLs=False)
print "Original peaks file: number indexed, error indexing = ", num,err
wksp=mtd['natrolite_orig']
lattice = wksp.sample().getOrientedLattice()
print "lattice, errors = ", lattice.a(),lattice.b(),lattice.c(),lattice.alpha(),lattice.beta(),lattice.gamma(),lattice.errora(),lattice.errorb(),lattice.errorc(),lattice.erroralpha(),lattice.errorbeta(),lattice.errorgamma()
print "U = ",lattice.getU()
print "B = ", lattice.getB()

#Levenberg-Marquardt optimization
start_time = time.time()
SCDCalibratePanels(PeakWorkspace='natrolite',
    useL0=True, usetimeOffset=False,usePanelWidth=True, usePanelHeight=True, usePanelPosition=True, usePanelOrientation=False, RotateCenters=True,
    AllowSampleShift=False, DetCalFilename='mandiLM.DetCal', ResultWorkspace='LMResults',
    ColWorkspace='LMCol', RowWorkspace='LMRow', TofWorkspace='LMTof', NumIterations=20)
print "minutes:  ", str((time.time()-start_time)/60)
num,err=IndexPeaks(PeaksWorkspace='natrolite',RoundHKLs=False)
print "MANTID optimized peaks file: number indexed, error indexing = ", num,err
wksp=mtd['natrolite']
lattice = wksp.sample().getOrientedLattice()
print "lattice, errors = ", lattice.a(),lattice.b(),lattice.c(),lattice.alpha(),lattice.beta(),lattice.gamma(),lattice.errora(),lattice.errorb(),lattice.errorc(),lattice.erroralpha(),lattice.errorbeta(),lattice.errorgamma()
print "U = ",lattice.getU()
print "B = ", lattice.getB()

#Simplex optimization
start_time = time.time()
SCDCalibratePanels(PeakWorkspace='natrolite_Simplex',
    useL0=True, usetimeOffset=False,usePanelWidth=True, usePanelHeight=True, usePanelPosition=True, usePanelOrientation=False, RotateCenters=True,
    AllowSampleShift=False, DetCalFilename='mandiSimplex.DetCal', ResultWorkspace='SimplexResults',
    ColWorkspace='SimplexCol', RowWorkspace='SimplexRow', TofWorkspace='SimplexTof', NumIterations=100,Minimizer='Simplex')
print "minutes:  ", str((time.time()-start_time)/60)
num,err=IndexPeaks(PeaksWorkspace='natrolite_Simplex',RoundHKLs=False)
print "MANTID Simplex optimized peaks file: number indexed, error indexing = ", num,err
wksp=mtd['natrolite_Simplex']
lattice = wksp.sample().getOrientedLattice()
print "lattice, errors = ", lattice.a(),lattice.b(),lattice.c(),lattice.alpha(),lattice.beta(),lattice.gamma(),lattice.errora(),lattice.errorb(),lattice.errorc(),lattice.erroralpha(),lattice.errorbeta(),lattice.errorgamma()
print "U = ",lattice.getU()
print "B = ", lattice.getB()

#Isaw Levenberg-Marquardt optimization
LoadIsawPeaks(Filename='natrolite_Niggli_0.12_2016.peaks', OutputWorkspace='natrolite_ISAW')
LoadIsawUB(InputWorkspace='natrolite_ISAW', Filename='natrolite_Niggli.mat')
num,err=IndexPeaks(PeaksWorkspace='natrolite_ISAW',RoundHKLs=False)
print "ISAW optimized peaks file: number indexed, error indexing = ", num,err
wksp=mtd['natrolite_ISAW']
lattice = wksp.sample().getOrientedLattice()
print "lattice, errors = ", lattice.a(),lattice.b(),lattice.c(),lattice.alpha(),lattice.beta(),lattice.gamma(),lattice.errora(),lattice.errorb(),lattice.errorc(),lattice.erroralpha(),lattice.errorbeta(),lattice.errorgamma()
print "U = ",lattice.getU()
print "B = ", lattice.getB()
print lattice.getB()
```

Fixes #15644 

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by wdzhou** [comment]
I tested this PR with the script given above.  The result looks reasonable. And thus the requirement of this PR/issue is met.

:shipit:

**Comment by wdzhou** [comment]
@VickieLynch will release note be change?

**Comment by VickieLynch** [comment]
@wdzhou I did commit a mention in release notes

File: Framework/Crystal/src/SCDCalibratePanels.cpp [hunk 1]
```diff
@@ -478,49 +488,48 @@ bool GoodStart(const PeaksWorkspace_sptr &peaksWs, double a, double b, double c,
                double alpha, double beta, double gamma, double tolerance) {
   // put together a list of indexed peaks
   std::vector<V3D> hkl;
-  hkl.reserve(peaksWs->getNumberPeaks());
+  int nPeaks = peaksWs->getNumberPeaks();
+  hkl.reserve(nPeaks);
   std::vector<V3D> qVecs;
-  qVecs.reserve(peaksWs->getNumberPeaks());
-  for (int i = 0; i < peaksWs->getNumberPeaks(); i++) {
+  qVecs.reserve(nPeaks);
```
Comment (FedeMPouzols) [inline Framework/Crystal/src/SCDCalibratePanels.cpp]: `resize()` and then assignments can be faster than `reserve()` and then push_backs. But I don't think it makes any difference in this case, only when you have large amounts of items.

**Comment by FedeMPouzols** [comment]
The release note looks good, and this looks like a great improvement. 
I imagine Leighton will be very happy about it.
It would be good to have a unit/system test for mandi data and calibrate-SCD in the future, so this is well protected. Probably the test script given in this PR can be used for a test?

**Pull request merged**